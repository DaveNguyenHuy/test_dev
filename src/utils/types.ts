export interface Movie {
  id: string;
  title: string;
  url: string;
  shared_by: string;
  description: string;
}

export interface MovieContextData {
  movies: Movie[];
  updateMovies: (movies: Movie[]) => void;
}
