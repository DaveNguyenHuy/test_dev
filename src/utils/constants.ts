export const ROUTES = {
  home: '/',
  share: '/share'
};

export const YouTubeGetID = (url: string) => {
  let ID: string | string[] = '';
  const parsedUrl = url.replace(/(>|<)/gi, '').split(/(vi\/|v=|\/v\/|youtu\.be\/|\/embed\/)/);
  if (parsedUrl[2] !== undefined) {
    return parsedUrl[2].split(/[^0-9a-z_-]/i)[0];
  }
  return ID;
};
