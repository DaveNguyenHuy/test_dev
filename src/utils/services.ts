import { collection, getDocs, query } from '@firebase/firestore';
import { db } from './firebase';

const moviesCollection = collection(db, 'movies');

export const fetchMovies = async () => {
  try {
    const q = query(moviesCollection);
    const querySnapshot = await getDocs(q);
    const data: any = [];
    querySnapshot.forEach((doc) => {
      data.push({
        id: doc.id,
        ...doc.data()
      });
    });
    return data;
  } catch (error) {
    return [];
  }
};
