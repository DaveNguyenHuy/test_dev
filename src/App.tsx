import { createContext, useState } from 'react';
import { Routes, Route } from 'react-router-dom';
import { Layout } from './components/Layout';

import HomePage from './pages/Home';
import NotFound from './pages/NotFound';
import ShareVideoPage from './pages/ShareVideo';
import { Movie, MovieContextData } from './utils/types';

export const MoviesContext = createContext<MovieContextData>({
  movies: [],
  updateMovies: () => null
});

function App() {
  const [movies, setMovies] = useState<Movie[]>([]);

  const updateMovies = (data: Movie[]) => {
    setMovies(data);
  };

  return (
    <MoviesContext.Provider value={{ movies, updateMovies }}>
      <Layout>
        <Routes>
          <Route path="/" element={<HomePage />} />
          <Route path="/share" element={<ShareVideoPage />} />
          <Route path="/*" element={<NotFound />} />
        </Routes>
      </Layout>
    </MoviesContext.Provider>
  );
}

export default App;
