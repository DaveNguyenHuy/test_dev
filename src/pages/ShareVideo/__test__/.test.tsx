import { render, screen, fireEvent } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { waitFor } from '@testing-library/dom';
import ShareVideoPage from '..';
import { MoviesContext } from '../../../App';
import { Movie } from '../../../utils/types';
import fetchMock from 'fetch-mock-jest';

jest.spyOn(window, 'alert').mockImplementation(() => {});

process.env.REACT_APP_YOUTUBE_API_KEY = 'api_key';

test('share video page should render correct', async () => {
  let movies: Movie[] = [];
  const updateMovies: any = jest.fn((data) => {
    movies = data;
  });

  const youtubeUrl = 'https://www.youtube.com/watch?v=Q5fd95h4HC';
  fetchMock.get(
    `https://www.googleapis.com/youtube/v3/videos?part=snippet&key=${process.env.REACT_APP_YOUTUBE_API_KEY}&id=Q5fd95h4HC`,
    {
      items: [
        {
          id: 'gtMV-baxuz0',
          snippet: {
            title: 'test video',
            description: 'description'
          }
        }
      ]
    }
  );

  const { container, getByRole } = render(
    <BrowserRouter>
      <MoviesContext.Provider
        value={{
          movies,
          updateMovies,
          appLoading: false,
          currentUser: { email: 'test@gmail.com' }
        }}
      >
        <ShareVideoPage />
      </MoviesContext.Provider>
    </BrowserRouter>
  );

  expect(container.firstChild).toMatchSnapshot();

  const submitButton = getByRole('button');

  fireEvent.change(screen.getByPlaceholderText('youtube url'), {
    target: { value: youtubeUrl }
  });
  fireEvent.click(submitButton);

  expect(submitButton).toHaveTextContent(/loading.../i);
  await waitFor(() => {});
  expect(submitButton).toHaveTextContent(/share/i);

  fireEvent.click(submitButton);

  await waitFor(() => {});
  expect(window.alert).toBeCalledWith('share movie success');
});
