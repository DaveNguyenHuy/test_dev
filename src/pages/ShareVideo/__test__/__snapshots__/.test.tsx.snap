// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`share video page should render correct 1`] = `
<div
  class="pt-24"
>
  <form
    class="mx-auto max-w-xl bg-white border-gray-500 relative border-2 px-8 py-12 rounded-md space-y-4 flex flex-col justify-end"
  >
    <h2
      class="rounded-md bg-white px-1 text-2xl absolute -top-5 left-4"
    >
      Share a Youtube Movie
    </h2>
    <div
      class="w-full space-x-2 flex"
    >
      <label
        class="text-lg mt-1"
      >
        Youtube URL:
      </label>
      <div
        class="space-y-6 flex-1"
      >
        <input
          class="focus:outline-none rounded-sm p-2 border border-gray-500 w-full"
          name="url"
          placeholder="youtube url"
          required=""
          value=""
        />
        <button
          class="p-1.5 border bg-blue-600 text-white rounded-md w-full"
        >
          Share
        </button>
      </div>
    </div>
  </form>
</div>
`;
